modules/ROOT/edits/dataweave-type-system.adoc

```diff
@@ -0,0 +1,357 @@
+
+For example, the `isNumeric` function from dw::core::Strings has the signature `isNumeric(String): Boolean` (receives a string and returns if that string is numeric) but it is also defined as `isNumeric(Null): Boolean`. +
+
+So if you call the function `isNumeric` with a null (`isNumeric(null)`) this does not throw an exception. At runtime, the function dispatching algorithm selects the correct function based on the actual types of the values used. The function returns false because of the definition of `isNumeric` for null as the argument.
+
+== Composite Types
+
+Composite types contains other values inside it. These composite types are Array, Object and Function.
+
+=== Array Type
+
+An Array is of type `Array&lt;T&gt;` where `T` is a type parameter that defines the type of the elements inside the Array. For example, the syntax to define a variable with the type Array of `Number` is:
+
+----
+var idsList: Array<Number> = [1, 22, 333, 4444]
+----
+
+The variable `idsList` only passes the type checking phase when the array assigned to it only contains Numbers.
+
+=== Object Type
+
+In order to define an object type there are two ways, either `Object` or `{}`, they both define an open object and don't specify any constraint to their key value pairs. Open and close objects are later defined in this section.
+
+The syntax to define an object type with a set of constraints for its key value pairs, is very similar to defining an object value:
+
+----
+{
+    keyName @(attrName: AttrType): valueType,
+    ...
+}
+----
+
+For example, to define a user type with `firstName` String, `lastName` String and `age` Number, the syntax is:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+type User = {
+    firstName: String,
+    lastName: String,
+    age: Number
+}
+----
+
+You can apply some modifiers to each key value pair to determine if a field is always present or if the field can be repeated or not.
+
+For repeated fields use `*`. For example, the following type allows to have multiple `lastName` fields:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+{
+    firstName: String,
+    lastName*: String,
+    age: Number
+}
+----
+
+For conditional fields use `?`. For example, the following type allows the age field to be present or not:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+{
+    firstName: String,
+    lastName*: String,
+    age?: Number
+}
+----
+
+==== Close and Open Objects
+
+Close object means that the type only accepts an object value if there are no additional fields in the object,
+just those key values specified by the type.
+
+Open object only put constraints on the fields that are declared in the type, and if they are all verified, then the type accepts the value. The type accepts objects that have other fields apart of the ones mentioned in the declaration. All object types are open unless specified.
```

Comment:
Same question about rewording "Open object only put constraints on ..." Could it be "An open object only puts constraints on..."

Reply:
also need to say what "they" refers to in the phrase "if they are all verified" explain what is verified.

Reply:
and "object types" -> "`Object` types"